### ISSUE TITLE
GPX: Show distance on Speed tab, if speed was calculated by OsmAnd.

### ISSUE THREAD

**Issue by dmpr0** [osmandapp/Osmand#3822]
Since yesterday we can calculate speed for GPX tracks if they don't have it. 
Now on Speed tab, missed Distance line (it was showing on Overview) and Time moving, if we could calculate it.
https://www.dropbox.com/s/d33ql0cuumeh311/GPX_Speed.png?dl=0
https://www.dropbox.com/s/lghumv7vd25knfb/GPX_Overview.png?dl=0

**Comment by njohnston** [comment]
From looking at [TrackSegmentFragment.java](https://github.com/osmandapp/Osmand/blob/master/OsmAnd/src/net/osmand/plus/myplaces/TrackSegmentFragment.java#L1192), the speed tab is apparently intended to show the total *moving* distance (`analysis.totalDistanceMoving`) rather than the distance (`analysis.totalDistance`) displayed on the overview tab.

For GPX files without speed, `totalDistanceMoving` is never set. From [GPXUtilities.java](https://github.com/osmandapp/Osmand/blob/master/OsmAnd/src/net/osmand/plus/GPXUtilities.java#L493):

```java
if ((speed > 0) && (calculations[0] > 0.1 / 1000f * (point.time - prev.time)) && point.time != 0 && prev.time != 0) {
  timeMoving = timeMoving + (point.time - prev.time);
  totalDistanceMoving += calculations[0];
}

// ...

//Last resorz: Derive speed values from displacment if track does not originally contain speed
if (!hasSpeedInTrack && speed == 0 && timeDiff > 0) {
  speed = calculations[0] / timeDiff;
}
```

In my pull request to add support for calculating speed (#3766), the `if (!hasSpeedInTrack ...` test was *before* the code to set `timeMoving` and `totalDistanceMoving`, so `speed` could be > 0 and then `timeMoving` and `totalDistanceMoving` could be updated.

@sonora: could we move the `if (!hasSpeedInTrack ...` code earlier? That would solve the problem, though it might have other implications.

**Comment by sonora** [comment]
Well, I am struggling with that.

I deliberately moved the code, because when I had originally introduced "timeMoving" and "totalDistanceMoving" some years ago, this data meant to explicitly report the values summed  **while the GPS chipset reported the device moving**.

Now I believe with our "speed calculated from displacement" fallback method, a device would seem to be moving all the time, due to omnipresent "displacement noise" in the position fix. Hence my intent was to explicitly suppress showing the "while moving" values, because they would not differ from the overall values. Or am I wrong?

**Comment by sonora** [comment]
Upon second thought, maybe you are right: We could shift the code (and by that report the device moving all the time), knowing that this can be an artifact if the speed curve is shown in red. But this wuld not be obvious, nor documented anywhere...  @dmpr0, @njohnston  What do you vote for?

**Comment by njohnston** [comment]
@sonora: thanks for your reply. I don't think "a device will _appear_ to be moving all the time" is quite accurate.

I moved the `if (!hasSpeedInTrack ...` code earlier, and with my test track, the time moving is reported as 1:43:54 compared to a total time of 1:44:58, so not moving for 1 minute and 4 seconds.

Adding debug output just before the `timeMoving` calculations shows several points where there has been insufficient movement (as per `calculations[0] > 0.1 / 1000f * (point.time - prev.time)`, shown as `movedEnough` below):

```
calculations[0] = 1.9645929, minDisplacement = 1.1, lat = 51.8997553, lon = -2.0729267, movedEnough = true
calculations[0] = 0.040895063, minDisplacement = 1.1, lat = 51.8997555, lon = -2.0729262, movedEnough = false
calculations[0] = 0.1141722, minDisplacement = 1.1, lat = 51.8997546, lon = -2.072927, movedEnough = false
calculations[0] = 0.92572623, minDisplacement = 1.0, lat = 51.8997629, lon = -2.072928, movedEnough = false
calculations[0] = 0.09536868, minDisplacement = 1.1, lat = 51.8997637, lon = -2.0729285, movedEnough = false
```

This corresponds to these points in the GPX file:

```xml
<trkpt lat="51.8997553" lon="-2.0729267">
  <ele>117.0</ele>
  <time>2017-05-01T10:42:25Z</time>
  <hdop>3.0</hdop>
</trkpt>
<trkpt lat="51.8997555" lon="-2.0729262">
  <ele>117.0</ele>
  <time>2017-05-01T10:42:36Z</time>
  <hdop>3.0</hdop>
</trkpt>
<trkpt lat="51.8997546" lon="-2.072927">
  <ele>117.0</ele>
  <time>2017-05-01T10:42:47Z</time>
  <hdop>3.0</hdop>
  </trkpt>
<trkpt lat="51.8997629" lon="-2.072928">
  <ele>115.0</ele>
  <time>2017-05-01T10:42:57Z</time>
  <hdop>3.0</hdop>
</trkpt>
<trkpt lat="51.8997637" lon="-2.0729285">
  <ele>115.0</ele>
  <time>2017-05-01T10:43:08Z</time>
  <hdop>9.0</hdop>
</trkpt>
```

In other words, the minimum displacement check still works here, though I think time not moving is slightly under-reported. Still, I think that's better than nothing. The not moving time correctly corresponds to a point on the track where I stopped, and like you say, every method has its own accuracy trade-offs.

I vote for moving the `if (!hasSpeedInTrack ...` code earlier, just before the minimum displacement check. I'm happy to move it, and I'd also like to clean up GPXUtilities.java. There are several typos in comments ('displacment', 'resorz'), and code which has been commented out for 6 months - 1.5 years.

**Comment by sonora** [comment]
Ah, ok, yes, I forgot I had a minimum displacement threshold anyway. So I guess you have a point there.

The ultimate test will be to simply use a track recorded strictly at rest, and check what the display shows.

**Comment by sonora** [comment]
Fixed via https://github.com/osmandapp/Osmand/commit/75a255f0ffadfd6a5a6aa690bc6d8988c147e18d.

**Comment by njohnston** [comment]
@sonora: thanks for the fix. Out of interest, I left my phone still on the windowsill and recorded a 7 minute track. OsmAnd reports the distance as 17m, presumably through accumulation of GPS error. OsmAnd does not insert `<speed>` into the track.

The calculated (displacement-based) speed is 0.5 km/h average, and 0.8 km/h maximum. Time moving is 0:52, and corrected distance is 8m. Not ideal, but acceptable, particularly as this is a rather contrived test.

**Comment by sonora** [comment]
Yes, I have just done a similar test with similar results, so I think we are good and can leave it as it is now. Thank you so much for the idea and solution!

In deed, we do not save `<speed>` to the GPX where the GPS chipset reports `speed = 0`, which is what I call "GPS chipset based motion detection" elsewhere and in some earlier discussions of the topic.
